src/ios-scroll.js

```diff
@@ -0,0 +1,513 @@
+/**
+ * Copyright 2015 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+/**
+ * The elapsed time cutoff that determines if a user intended to stop the
+ * synthetic scroll momentum.
+ * @const @private {number}
+ */
+const SCROLL_INTENTION_ELAPSED_CUTOFF = (1000 / 60) * 3;
+
+/**
+ * Manages scrolling of embedded AMP document in iOS, since iOS doesn't
+ * natively scroll embedded documents well. This "synthetic" scrolling keeps
+ * track of fixed position elements so that they may be scrolled in unison. It
+ * also keeps track of iframes so that they can be disabled during scrolling
+ * (so they don't start a native scroll during our synthetic scroll).
+ *
+ * This is a substantially modified version of Kinetic Scrolling.
+ */
+export class SyntheticScroll {
+
+  /**
+   * @param {HTMLBodyElement} body
+   * @param {number} scrollTop the body's current scrollTop
+   * @param {number} scrollHeight the body's current scrollHeight
+   */
+  constructor(scrollEl, scrollTop, scrollHeight) {
+    const doc = scrollEl.ownerDocument;
+    const body = doc.body;
+
+    /**
+     * @const @private {!Window}
+     */
+    this.win_ = doc.defaultView;
+
+    /**
+     * We store the style reference directly to speed things up.
+     * @const @private {!CSSStyleDeclaration}
+     **/
+    this.scrollStyle_ = scrollEl.style;
+
+    /**
+     * We track the iframes on the page so we may disable direct interaction
+     * during scrolling. This prevents them from causing native scrolling.
+     * When scrolling is done, they will be re-enabled.
+     * @const @private {!Array<!CSSStyleDeclaration>}
+     **/
+    this.iframeStyles_ = [];
+
+    /**
+     * A list of fixed position elements that we should scroll with the doc.
+     * @const @private {!Array<!CSSStyleDeclaration>}
+     */
+    this.fixedElementStyles_ = [];
+
+    /**
+     * A helper element that records the last position we scrolled to. This
+     * allows us to determine how much the user scrolled natively.
+     * @const @private {Element}
+     */
+    this.scrollMoveEl_ = doc.createElement('div');
```

Comment:
We are storing it's position in memory (`#scrollTop_`). This element is needed because the document can native scroll due to an iframe. In order to tell how far that scroll moved us, I use this element's location. Also, using this element allows us to scroll the body through the API (`this.scrollMoveEl_.scrollIntoView(true)`).